autotest/cpp/test_gdal_pixelfn.cpp

```diff
@@ -0,0 +1,262 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Project:  C++ Test Suite for GDAL/OGR
+// Purpose:  Test constant and builtin arguments for C++ pixel functions
+// Author:   Momtchil Momtchev <momtchil@momtchev.com>
+//
+///////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2022, Momtchil Momtchev <momtchil@momtchev.com>
+/*
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "gdal_unit_test.h"
+
+#include "cpl_string.h"
+#include "gdal_alg.h"
+#include "gdal_priv.h"
+#include "gdal.h"
+#include "../frmts/vrt/vrtdataset.h"
+
+#include <sstream>
+#include <string>
+#include <vector>
+
+CPLErr CustomPixelFuncWithMetadata( void **papoSources, int nSources, void *pData,
+                            int nXSize, int nYSize,
+                            GDALDataType eSrcType, GDALDataType eBufType,
+                            int nPixelSpace, int nLineSpace, CSLConstList papszArgs );
+CPLErr CustomPixelFunc( void **papoSources, int nSources, void *pData,
+                            int nXSize, int nYSize,
+                            GDALDataType eSrcType, GDALDataType eBufType,
+                            int nPixelSpace, int nLineSpace, CSLConstList papszArgs );
+CPLErr CustomPixelFuncNoArgs( void **papoSources, int nSources, void *pData,
+                            int nXSize, int nYSize,
+                            GDALDataType eSrcType, GDALDataType eBufType,
+                            int nPixelSpace, int nLineSpace);
+
+template<typename T>
+inline double
+GetSrcVal(const void* pSource, GDALDataType eSrcType, T ii)
```

Comment:
This is the duplicate of a function in ``pixelfunctions.cpp``.
Maybe we could move it to a common header ``pixelfunctions.hpp``.

Reply:
Yes, but in this case it is to be considered public API and to be properly referenced.

I also suggest that both the OSGeo and OpenJS foundations finally realize and come to terms with the situation, as accidently running down a pedestrian is obviously a problem, but finishing him off is a huge gamble and rarely has the intended result.

Reply:
> Yes, but in this case it is to be considered public API and to be properly referenced.

I don't think that we are forced to consider it a public API to to install the new ``pixelfunctions.hpp``.

Reply:
> to install the new pixelfunctions.hpp.

Actually we have the SRCVAL() macro in gdal.h thta does a similar job

Reply:
Yes,  I saw it, I was thinking of a way to merge both.
As this is performance-critical, I am also thinking of a single switch, performed once per pixel function call, and then using a lambda to cast each pixel.

Reply:
actually SRCVAL() is the historical way of doing things, but for internal pixel functions, I introduced the template in 9048844951bd8b31f179f709126a1c3fb0cd4cc1 (to fix a cppcheck warning)
More than this switch() if you're concerned about performance, the GDALCopyWords() function call is certainly much more heavy weight. The GDALCopyWord() template function (in gcore/gdal_priv_templates.hpp) could potentially be used.
But then you need to have a template on the source and target data types.

Reply:
Templated functions wins for every compiler:
https://gist.github.com/mmomtchev/ae0251f0d400db625b9d33069237a547

-O3 everywhere

g++ 9.3:
templated function: 520 ms
macro: 606 ms
function pointer: 628 ms
lambda: 786 ms

clang 12:
templated function: 512 ms
macro: 512 ms
function pointer: 608 ms
lambda: 822 ms

intel compiler 2022:
templated function: 469 ms
macro: 470 ms
function pointer: 630 ms
lambda: 785 ms